### ISSUE TITLE
Limit parameter is confusing

### ISSUE THREAD

**Issue by roysmith** [mwclient/mwclient#259]
I've got version 0.10.0.  Running this code:

```
import mwclient

site = mwclient.Site('en.wikipedia.org')
results = site.usercontributions('RoySmith-Mobile', limit=5)
print(len(list(results)))
```

prints:

```
91
```

Expected result: it should print "5".  As far as I can tell, the limit parameter is ignored.

**Comment by roysmith** [comment]
Sigh.  Did some more digging.  The limit parameter is just passed down to the API level where it really means the chunk size over which the List object iterates.  The point of a high-level API is to provide a higher level abstraction, not just expose all the low level details in a slightly different skin.

**Comment by danmichaelo** [comment]
For what it's worth, I agree that the parameter name is quite confusing! Wouldn't really mind changing it to `chunk_size`, except it's a breaking change, so the old parameter name would have to deprecated and kept for some time.

Also agree it's not very elegant having to do something like this:

```
import itertools
contribs = list(itertools.islice(site.usercontributions('RoySmith-Mobile'), 5))
```

or 

```
results_generator = site.usercontributions('RoySmith-Mobile')
results = [next(results_generator) for _ in range(5)]
```

just for a common operation like fetching a limited set of results. Thinking about it, one way to make the API a bit more fluent without breaking backwards compatibility, could be to add a new method on the `List` class, that would return the first N results:

```python
results = site.usercontributions('RoySmith-Mobile').take(5)
```

The method could look something like this:

```python
def take(self, limit):
    """Return a list of up to `limit` results"""
    return list(itertools.islice(self, limit))
```

**Comment by roysmith** [comment]
I'm unsure we want to start adding methods like that; we already have the general-purpose itertools library which does all of these sorts of things.

But, more than that, returning an iterator is good.  We certainly don't want to listify the results.  What if you looked at the contributions for [[User:User:Ser Amantio di Nicolao]], with [over 3 million edits](https://en.wikipedia.org/wiki/Wikipedia:List_of_Wikipedians_by_number_of_edits#1%E2%80%931000)?

I noticed this is called out in [a debian release note](https://packages.debian.org/sid/python3-mwclient) so I guess I'm not the first person to be bitten by this.

**Comment by danmichaelo** [comment]
Absolutely agree that iterators should be the standard. This would just be a convenience method, to use when requesting a small number of results, which I think is quite a common use case. But it would only be a minor convenience over using `itertools.islice` directly, so it might not be worth it.

I see that we already have mentioned `islice` at https://mwclient.readthedocs.io/en/latest/user/page-ops.html#listing-page-revisions , but it could perhaps be mentioned more places. Let me know if you have specific ideas for improving the documentation. Perhaps adding a warning to some of the docstrings would help?

**Comment by roysmith** [comment]
Looking through the code, I see listing.List already has [support for a max_items parameter](https://github.com/mwclient/mwclient/blob/7c097f2a947122c2470f44a4f153dc44f57c09f4/mwclient/listing.py#L18) which looks like it does the right thing.  There's also a mysterious ["# NOTE: Fix limit"](https://github.com/mwclient/mwclient/blob/7c097f2a947122c2470f44a4f153dc44f57c09f4/mwclient/listing.py#L20) comment.  I haven't dug far enough back in the history, but I'm assuming these are related.

So, maybe just expose max_items at the Site.usercontributions() level?

As much as I argued against duplicating the islice() functionality inside mwclient, there is at least the self-documenting feature.  If I was a lazy programmer and didn't read all the documentation before doing anything (yeah, right), at least I would see that the method had both a "max_items" and a "limit" and think to myself, "Oh, that's weird, how do these two differ from each other?", and go digging a bit deeper.

And, to be honest

`site.usercontributions(user, max_items=n)`

is easier to read than

`islice(site.usercontributions(user), n)`

**Comment by roysmith** [comment]
Ooops, sorry, didn't mean to close that.  Clicked the wrong button.

**Comment by danmichaelo** [comment]
Interesting, I have actually never noticed the `max_items` parameter before :D I'm open to the idea of exposing that on the various list methods. It's not ideal, but as you say, it might cause some users to stop, think and read the documentation when they see both `limit` and `max_items` :) Would you be interested in creating a PR for that?

**Comment by AdamWill** [comment]
@RheingoldRiver and I looked into this and it's actually quite a big problem. There's a lot of places in the code that take a kwarg `limit` and pass it through to one of the `listing` classes, and several of them are written and documented as if `limit` worked the way Roy thought it did, e.g. `Site.random()` - which says "limit specifies the number of random articles retrieved", but it definitely doesn't - and `Page.revisions()` - which says "limit (int): The maximum number of revisions to return per request", but again, it's definitely not.

It will take a big commit to fix all this, but at the same time, it does seem like it's a pretty big issue and we should fix it.

**Comment by AdamWill** [comment]
Been digging into this further today...couple more notes: this has worked the same way ever since the first commit to the repo that contained any code - e93b7290fc96b5218408930275f93ba4170dd07c - and Bryan was aware of it. There was a note in `README.txt` in that commit that says "Their limit parameter is only an indication of the number of items in one chunk. It is not the total limit. Doing list(generator(limit = limit)) will return ALL items of generator, and not be limitted by the limit value." We actually have kept that note, but it's now buried in `docs/source/user/implementation-notes.rst`.

Given this, I'm hesitant to go with my first idea of changing how the high-level functions behave. Instead I'm working on a PR that consistently adds `chunk_size` and `max_items` args to everything that currently has a `limit` arg, and makes `limit` a deprecated synonym for `chunk_size`.

**Comment by marcfrederick** [comment]
The `limit` parameter was replaced with `max_items` and `max_chunk_size` in #343 as described in @AdamWill's previous comment. The change is part of v0.11.0.

**Comment by roysmith** [comment]
Hey, just want to say I'm glad folks are still maintaining mwclient, because I'm still using it!  And I think the fix outlined in #343  was the right way to go.
